src/test/osd/TestOSDMap.cc

```diff
@@ -2073,6 +2073,126 @@ TEST_P(OSDMapTest, BUG_51842) {
     }
 }
 
+TEST_F(OSDMapTest, blocklisting) {
+  set_up_map(6); //whatever
+
+  // I pulled the first two ranges and their start/end points from
+  // https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing#CIDR_notation
+  string range_addrs[] = {"198.51.100.0/22", "2001:db8::/48", "3001:db8::/72"};
+  string ip_addrs[] = {"198.51.100.14", "198.51.100.0", "198.51.103.255",
+    "2001:db8:0:0:0:0:0:0", "2001:db8:0:0:0:0001:ffff:ffff",
+    "2001:db8:0:ffff:ffff:ffff:ffff:ffff",
+    "3001:db8:0:0:0:0:0:0", "3001:db8:0:0:0:0001:ffff:ffff",
+    "3001:db8:0:0:00ff:ffff:ffff:ffff", };
+  string unblocked_ip_addrs[] = { "0.0.0.0", "1.1.1.1", "192.168.1.1",
+    "198.51.99.255", "198.51.104.0",
+    "2001:db7:ffff:ffff:ffff:ffff:ffff:ffff", "2001:db8:0001::",
+    "3001:db7:ffff:ffff:ffff:ffff:ffff:ffff", "3001:db8:0:0:0100::"
+  };
+  OSDMap::Incremental new_blocklist_inc(osdmap.get_epoch() + 1);
+  for (const auto& a : ip_addrs) {
+    entity_addr_t addr;
+    addr.parse(a);
+    addr.set_type(entity_addr_t::TYPE_LEGACY);
+    new_blocklist_inc.new_blocklist[addr] = ceph_clock_now();
+  }
+  osdmap.apply_incremental(new_blocklist_inc);
+
+  for (const auto& a: ip_addrs) {
+    entity_addr_t addr;
+    addr.parse(a);
+    addr.set_type(entity_addr_t::TYPE_LEGACY);
+    ASSERT_TRUE(osdmap.is_blocklisted(addr, g_ceph_context));
+  }
+  for (const auto& a: unblocked_ip_addrs) {
+    entity_addr_t addr;
+    addr.parse(a);
+    addr.set_type(entity_addr_t::TYPE_LEGACY);
+    ASSERT_FALSE(osdmap.is_blocklisted(addr, g_ceph_context));
+  }
```

Comment:
This test should stop here. I'd like to see this broken up into separate tests that do:

* Test current blocklist behavior 1 address matching among many blocklisted addresses
* Test current blocklist behavior 1 address not matching
* Test current blocklist behavior with IP w/o nonce/port
* Test removing a blocklist item
* Test different synonymous CIDR ranges like: `2001::fffe/128` and `2001::0:fffe/128`. Ditto for ipv6 to match against.
* Test CIDR edge-cases (/128 and /0)

Reply:
I split the tests up some (thought not all the way) and I think we have all the edge cases you've mentioned here now.